internal/exec/stages/disks/luks.go

```diff
@@ -531,3 +570,149 @@ func (d LuksDump) hasFlag(flag string) bool {
 	}
 	return false
 }
+
+// collect the cex card domains xx.xxxx
+func getAllCexOnlineAPQNs() (string, error) {
+	apDir := "/sys/bus/ap/devices/"
+	uEvent := "/uevent"
+	var ret []string
+	// cards details are in form of link file points to card directory.
+	event, _ := os.ReadDir(apDir)
+	if len(event) == 0 {
+		return "", fmt.Errorf("adjunct processor not found")
+	}
+	// match the regexp for pattern xx.xxxx
+	for _, dir := range event {
+		match := cexRegx.FindStringSubmatch(dir.Name())
+		// if a match is found, read the uevent from that APQN directory
+		// to verify whether it corresponds to a CCA controller. This is
+		// intended for potential future use with an EP11 controller.
+		if match != nil {
+			filePath := apDir + strings.Join(match, "") + uEvent
+			ctrl, err := getCryptoController(filePath)
+			if err != nil {
+				return "", fmt.Errorf("failed to parse uevent file: %w", err)
+			}
+			for key := range ctrl {
+				if ctrl[key] == "cca" {
+					ret = append(ret, match[0])
+				}
+			}
+			if len(ret) == 0 {
+				return "", fmt.Errorf("cannot find the cca ap controller")
+			}
+		}
+	}
+	return strings.Join(ret, ","), nil
+}
+
+// Find the AP (Adjunct Processor) mode from uevent.
+// It may have CCA or EP11 CEX Controllers.
+func getCryptoController(filePath string) (map[string]string, error) {
+	apPairs := make(map[string]string)
+	data, err := os.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read uevent file %q: %w", filePath, err)
+	}
+	content := string(data)
+	for _, line := range strings.Split(content, "\n") {
+		apParts := strings.SplitN(line, "=", 2)
+		if len(apParts) == 2 {
+			key := strings.TrimSpace(apParts[0])
+			value := strings.TrimSpace(apParts[1])
+			apPairs[key] = value
+		}
+	}
+	return apPairs, nil
+}
+
+// Arguments specific to zkey generation
+func (s *stage) zkeySecGenArgs(luks types.Luks) []string {
+	ret := []string{
+		"generate",
+		"--name",
+		"ignition-luks-" + luks.Name,
+		"--xts",
+		"--description",
+		"ignition-luks-" + luks.Name,
+	}
+	// Verify the KeyType for securekey
+	if !util.NilOrEmpty(luks.Cex.KeyType) {
+		ret = append(ret, "--key-type", *luks.Cex.KeyType)
+	} else {
+		ret = append(ret, "--key-type", "CCA-AESCIPHER")
+	}
+	return ret
+}
+
+// Secure key Generation using zkey.
```

Comment:
Probably worth mentioning here that the keys will be generated in the repository.

Reply:
Will correct it.